### PR TITLE
Implement `flatiter.__iter__()`

### PR DESCRIPTION
**Pull request by takagi**
This PR implements `flatiter.__iter__()` special method.

### CODE REVIEW

File: tests/cupy_tests/indexing_tests/test_iterate.py [hunk 1]
```diff
@@ -15,6 +15,10 @@ def test_base(self):
         a = cupy.zeros((2, 3, 4))
         assert a.flat.base is a
 
+    def test_iter(self):
+        it = cupy.zeros((2, 3, 4)).flat
+        assert it.__iter__() is it
```
Comment (asi1024) [inline tests/cupy_tests/indexing_tests/test_iterate.py:20]: How about writing as follows to check the compatibility to NumPy?
```py
for xp in (numpy, cupy):
    it = xp.zeros((2, 3, 4)).flat
    assert iter(it) is it
```
I think the other tests should be written as this format too, but It's OK to fix in another PR.

Comment (takagi) [reply tests/cupy_tests/indexing_tests/test_iterate.py]: I've fixed it as such and, okay, I will change the other tests too!

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**